kernel/src/expressions/literal_expression_transform.rs

```diff
@@ -0,0 +1,526 @@
+#[derive(Debug)]
+pub(crate) struct LiteralExpressionTransform<'a, T: Iterator<Item = &'a Scalar>> {
+    /// Leaf values to insert in schema order.
+    scalars: T,
+    /// A stack of built Expressions. After visiting children, we pop them off to
+    /// build the parent container, then push the parent back on.
+    stack: Vec<Expression>,
+    /// Since schema transforms are infallible we keep track of errors here
+    error: Result<(), Error>,
+}
+
+/// Any error for [`LiteralExpressionTransform`]
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    /// Schema mismatch error
+    #[error("Schema error: {0}")]
+    Schema(String),
+
+    /// Insufficient number of scalars (too many) to create a single-row expression
+    #[error("Too many scalars given for literal expression transform")]
+    ExcessScalars,
+
+    /// Insufficient number of scalars (too few) to create a single-row expression
+    #[error("Too few scalars given for literal expression transform")]
+    InsufficientScalars,
+
+    /// Empty expression stack after performing the transform
+    #[error("No Expression was created after performing the transform")]
+    EmptyStack,
+
+    /// Unsupported operation
+    #[error("Unsupported operation: {0}")]
+    Unsupported(String),
+}
+
+impl<'a, I: Iterator<Item = &'a Scalar>> LiteralExpressionTransform<'a, I> {
+    pub(crate) fn new(scalars: impl IntoIterator<Item = &'a Scalar, IntoIter = I>) -> Self {
+        Self {
+            scalars: scalars.into_iter(),
+            stack: Vec::new(),
+            error: Ok(()),
+        }
+    }
+
+    /// return the Expression we just built (or propagate Error). the top of `stack` should be our
+    /// final Expression
+    pub(crate) fn try_into_expr(mut self) -> Result<Expression, Error> {
+        self.error?;
+
+        if self.scalars.next().is_some() {
+            return Err(Error::ExcessScalars);
+        }
+
+        match self.stack.pop() {
+            Some(expr) => Ok(expr),
+            None => Err(Error::EmptyStack),
+        }
+    }
+
+    fn set_error(&mut self, error: Error) {
+        if let Err(e) = mem::replace(&mut self.error, Err(error)) {
+            debug!("Overwriting error that was already set: {e}");
+        }
+    }
+
+    fn check_error<T, E: Into<Error>>(&mut self, result: Result<T, E>) -> Option<T> {
+        match result {
+            Ok(val) => Some(val),
+            Err(err) => {
+                self.set_error(err.into());
+                None
+            }
+        }
+    }
+}
+
+impl<'a, T: Iterator<Item = &'a Scalar>> SchemaTransform<'a> for LiteralExpressionTransform<'a, T> {
+    fn transform_primitive(
+        &mut self,
+        prim_type: &'a PrimitiveType,
+    ) -> Option<Cow<'a, PrimitiveType>> {
+        // first always check error to terminate early if possible
+        self.error.as_ref().ok()?;
+
+        let next = self.scalars.next();
+        let scalar = self.check_error(next.ok_or(Error::InsufficientScalars))?;
```

Comment:
nit: this is the only remaining caller of `check_error`... inline it to simplify the logic?
```suggestion
        let Some(scalar) = self.scalars.next() else {
            self.set_error(Error::InsufficientScalars);
            return None;
        };
```

Reply:
oh good call - missed that we were only using once now!